### PR TITLE
[java] Fix false positive with UseStringBufferForStringAppendsRule

### LINKED ISSUE #222: [java] UseStringBufferForStringAppends: False Positive with ternary operator
**Rule Set:** java-optimizations/UseStringBufferForStringAppends

**Description:**
False Positive, similar to https://sourceforge.net/p/pmd/bugs/1340/, but with flipped condition and arguments.

**Code Sample demonstrating the issue:**

```
public class UseStringBuffer {
    public void foo() {
        String country = request.getParameter("country");
        country = (country != null) ? country : "USA";
    }
}
```

**Running PMD through:** CLI

### PR DESCRIPTION
**Pull request by Oowekyala**
Fixes #222 
**PR Description:** This fixes a bug which caused a false positive from UseStringBufferForStringAppendsRule on a snippet such as
```java
String country = request.getParameter("country");
country = (country != null) ? country : "USA";
```

The reason the rule trigerred a false positive in that case and not when inverting the ternary's parameters (`(country != null) ? "USA" : country;`) is that both parameters are not symmetrical (see [java spec](https://docs.oracle.com/javase/specs/jls/se8/html/jls-15.html#jls-15.25) here). The first is an `Expression` while the other is a `PrimaryExpression`. The first being an `Expression` actually adds one parent to the name occurence. For instance for the following statement:
```java
country = (country != null) ? country : country;
```
the AST of the right-hand side expression looks like this : 
```
ConditionalExpression
    PrimaryExpression
    Expression
        PrimaryExpression
            PrimaryPrefix
                Name @Image="country"
    PrimaryExpression
        PrimaryPrefix
            Name @Image="country"
```

Since in the visitor the name occurence is not reported if its *third* parent is the conditional expression, it did not catch the fact that when used as first argument, the name occurence actually has the conditional expression as its *fourth* parent.

**Solution:** Check for the third and the fourth parent of the name occurence

### CODE REVIEW

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/optimizations/xml/UseStringBufferForStringAppends.xml [hunk 1]
```diff
@@ -160,6 +160,7 @@ public class UseStringBuffer {
     public void foo2() {
         String country = request.getParameter("country");
         country = (country == null) ? "USA" : country;
+        country = (country != null) ? country : "USA";
```
Comment (jsotuyod) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/optimizations/xml/UseStringBufferForStringAppends.xml]: I'd rather have a new test case for this scenario, as seen they are independent (2 separate issues were reported and fixed on their own).

**Comment by jsotuyod** [comment]
Also, a piece of advise. Create feature branches for each changeset. That way you may work on different issues / features in parallel while waiting for reviews. You can read a little about it here: https://git-scm.com/book/en/v2/Git-Branching-Branching-Workflows#_topic_branch

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/optimizations/xml/UseStringBufferForStringAppends.xml (hunk 1 above)
Comment (Oowekyala) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/optimizations/xml/UseStringBufferForStringAppends.xml]: Okay that's fixed

**Comment by Oowekyala** [comment]
@jsotuyod Thanks, that's sound advice. I'll keep that in mind next time :)

**Comment by jsotuyod** [comment]
Thanks again for the PR! This is now merged and will be included in all future PMD releases.

**Pull request merged**